File: src/net/eduard/permissions/manager/PermissionsManager.java
package net.eduard.permissions.manager;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.bukkit.entity.Player;

import net.eduard.api.lib.game.FakePlayer;
import net.eduard.api.lib.storage.Storable;

/**
 * Controlador de Permissões e Grupos dos Jogadores
 * 
 * @author Eduard-PC
 *
 */
public class PermissionsManager {

	@Storable.StorageAttributes(reference = true)
	private PermissionsGroup groupDefault;

	private List<PermissionsGroup> groups = new ArrayList<>();

	private List<PermissionsPlayer> players = new ArrayList<>();

	public PermissionsPlayer getPlayer(Player player) {
		return getPlayer(new FakePlayer(player));
	}
	public void removeGroup(PermissionsGroup group) {
		for (PermissionsPlayer user : players) {
			user.getGroups().remove(group);
		}
		groups.remove(group);
	}
	public PermissionsGroup getGroup(String name) {
		for (PermissionsGroup group : groups) {
			if (group.getName().equalsIgnoreCase(name)) {
				return group;
			}
		}
		return null;
	}

	public PermissionsGroup createGroup(String name, String prefix, String suffix, String... permissoes) {
		PermissionsGroup group = new PermissionsGroup();
		group.setName(name);
		group.getPermissions().addAll(Arrays.asList(permissoes));
		groups.add(group);
		return group;
	}

	public PermissionsPlayer getPlayer(FakePlayer player) {
		for (PermissionsPlayer playerRegistred : players) {
			if (playerRegistred.getPlayer().equals(player)) {
				return playerRegistred;
			}
		}
		return null;

	}


	public List<PermissionsPlayer> getPlayers() {
		return players;
	}

	public void setPlayers(List<PermissionsPlayer> players) {
		this.players = players;
	}

	public List<PermissionsGroup> getGroups() {
		return groups;
	}

	public void setGroups(List<PermissionsGroup> groups) {
		this.groups = groups;
	}

	public PermissionsGroup getGroupDefault() {
		return groupDefault;
	}

	public void setGroupDefault(PermissionsGroup groupDefault) {
		this.groupDefault = groupDefault;
	}

	public PermissionsPlayer createPlayer(FakePlayer fakePlayer, String prefix, String suffix, String... permissions) {

		PermissionsPlayer player = new PermissionsPlayer();
		player.setPlayer(fakePlayer);
		player.getGroups().add(groupDefault);
		player.getPermissions().addAll(Arrays.asList(permissions));
		player.setPrefix(prefix);
		player.setSuffix(suffix);
		players.add(player);
		return player;

	}

}

File: src/main/java/net/eduard/permissions/core/PermissionsManager.kt
package net.eduard.permissions.core

import net.eduard.api.lib.command.PlayerOffline
import net.eduard.api.lib.plugin.IPlugin
import net.eduard.permissions.api.PermGroup
import net.eduard.permissions.api.PermissionsAPI

/**
 * Controlador de Permissões e Grupos dos Jogadores
 *
 * @author Eduard
 */
class PermissionsManager : PermissionsAPI{

    fun setDefaultGroup(group : PermissionsGroup) {
        groupDefault = group
    }
    override lateinit var groupDefault : PermGroup

    @Transient
    override var players = mutableMapOf<PlayerOffline, PermissionPlayer>()

    override fun registerGroup(group: PermGroup) {
        groups[group.name.toLowerCase()] = group as PermissionsGroup
    }

    override lateinit var plugin: IPlugin


    override var groups = mutableMapOf<String, PermissionsGroup>()


    fun removeGroup(group: PermissionsGroup) {
        for (user in players.values) {
            user.groups.remove(group)
        }
        groups.remove(group.toString().toLowerCase())
    }

    override fun getGroup(name: String): PermissionsGroup {
        return groups[name.toLowerCase()]!!
    }

    override fun createGroup(
        name: String,
        prefix: String,
        suffix: String,
        vararg permissions: String
    ): PermissionsGroup {
        val group = PermissionsGroup()
        group.name = name
        for (perm in permissions){
            group.permissions[perm.toLowerCase()] = true
        }

        groups[group.name.toLowerCase()] = group
        return group
    }

    override fun getPlayer(player: PlayerOffline): PermissionPlayer {
        return players.getOrDefault(player, newAccount(player))
    }

    fun newAccount(offline: PlayerOffline): PermissionPlayer {
        val player = PermissionPlayer()
        player.player = offline
        player.groups.add(groupDefault as PermissionsGroup)
        players[offline] = player
        return player
    }


}